File: IFATCCompanion/Review/ReviewRequestManager.swift
import Foundation
import SwiftUI
import StoreKit

/// Coordinates App Store rating & review prompts, following Apple's guidance for
/// StoreKit's `requestReview` action:
///
/// * The system displays the prompt **at most three times per 365 days**, and may
///   decline to show it at all — a call is a *request*, not a command.
/// * Apple recommends asking only after the user has engaged with the app and
///   completed something meaningful, never immediately on launch, and never in the
///   middle of a task.
///
/// On top of the system limit this manager adds a conservative app-side gate so a
/// prompt only ever appears at the two calm, positive moments the product allows —
/// **before the pilot's first ATC call of a session**, or **after a completed
/// flight** — never during a flight, never on a brand-new install, and no more
/// often than once every few months.
///
/// The manager only *decides* when to ask; presenting the sheet is a SwiftUI
/// view-environment action (`\.requestReview`), so a view observes `promptToken`
/// and performs the actual request (see `View.reviewPrompt(_:)`).
@MainActor
final class ReviewRequestManager: ObservableObject {

    /// The two low-workload windows in which a prompt is permitted. Mirrors the
    /// product rule: never mid-flight.
    enum Trigger {
        /// Connected and idle at the gate, before the first ATC call of the session.
        case beforeFirstCall
        /// Arrived and parked — the flight is complete.
        case afterFlightComplete
    }

    /// Incremented whenever an eligible moment asks for the prompt. A view bound to
    /// this value calls the StoreKit `requestReview` action in response.
    @Published private(set) var promptToken = 0

    // MARK: - Tuning

    /// Completed flights required before the app ever asks — keeps brand-new users
    /// and first-launch sessions out of the prompt entirely.
    private let minimumCompletedFlights = 3
    /// Minimum age of the install before the first ask. Defensive (a fresh install
    /// also has zero completed flights), and honours Apple's "not on first launch".
    private let minimumInstallAge: TimeInterval = 3 * 86_400
    /// Minimum spacing between two prompts. ~120 days keeps us well inside Apple's
    /// three-per-year system limit and avoids nagging.
    private let minimumInterval: TimeInterval = 120 * 86_400
    /// App-side mirror of Apple's own hard cap (three per rolling 365 days), so a
    /// run of tightly spaced eligible moments can never overshoot it.
    private let maximumPromptsPerYear = 3
    private let oneYear: TimeInterval = 365 * 86_400

    // MARK: - Dependencies

    private let defaults: UserDefaults
    private let now: () -> Date

    init(defaults: UserDefaults = .standard, now: @escaping () -> Date = Date.init) {
        self.defaults = defaults
        self.now = now
        // Stamp the install date the first time the manager ever runs, so the
        // install-age gate has a baseline.
        if installDate == nil { installDate = now() }
    }

    // MARK: - Recording engagement

    /// Record a fully completed flight (arrived and parked). Drives the engagement
    /// gate; call exactly once per completed flight.
    func recordFlightCompleted() {
        completedFlights += 1
    }

    // MARK: - Requesting

    /// Ask the system to present the rating prompt if — and only if — every app-side
    /// gate passes. Safe to call at either allowed moment; it self-limits and is a
    /// no-op when not eligible.
    func requestReviewIfAppropriate(_ trigger: Trigger) {
        guard isEligible else { return }
        recordPromptRequested()
        promptToken &+= 1
    }

    // MARK: - Eligibility

    /// Whether all app-side gates currently permit a prompt. The system applies its
    /// own three-per-year limit on top of this.
    var isEligible: Bool {
        guard completedFlights >= minimumCompletedFlights else { return false }
        if let installed = installDate,
           now().timeIntervalSince(installed) < minimumInstallAge { return false }
        if let last = lastPromptDate,
           now().timeIntervalSince(last) < minimumInterval { return false }
        if promptsInLastYear >= maximumPromptsPerYear { return false }
        return true
    }

    private func recordPromptRequested() {
        let t = now()
        lastPromptDate = t
        recentPromptDates = recentPromptDates.filter { t.timeIntervalSince($0) < oneYear } + [t]
    }

    private var promptsInLastYear: Int {
        let t = now()
        return recentPromptDates.filter { t.timeIntervalSince($0) < oneYear }.count
    }

    // MARK: - Persistence (UserDefaults)

    private enum Key {
        static let installDate = "review.installDate"
        static let completedFlights = "review.completedFlights"
        static let lastPromptDate = "review.lastPromptDate"
        static let recentPromptDates = "review.recentPromptDates"
    }

    private(set) var installDate: Date? {
        get { defaults.object(forKey: Key.installDate) as? Date }
        set { defaults.set(newValue, forKey: Key.installDate) }
    }

    private(set) var completedFlights: Int {
        get { defaults.integer(forKey: Key.completedFlights) }
        set { defaults.set(newValue, forKey: Key.completedFlights) }
    }

    private var lastPromptDate: Date? {
        get { defaults.object(forKey: Key.lastPromptDate) as? Date }
        set { defaults.set(newValue, forKey: Key.lastPromptDate) }
    }

    private var recentPromptDates: [Date] {
        get { (defaults.array(forKey: Key.recentPromptDates) as? [Date]) ?? [] }
        set { defaults.set(newValue, forKey: Key.recentPromptDates) }
    }
}

// MARK: - View integration

/// Presents the system App Store review prompt whenever the manager signals an
/// eligible moment. `requestReview` is a view-environment action, so the trigger
/// has to live in the view layer; the manager only decides *when*.
private struct ReviewPromptModifier: ViewModifier {
    @ObservedObject var manager: ReviewRequestManager
    @Environment(\.requestReview) private var requestReview

    func body(content: Content) -> some View {
        content.onChange(of: manager.promptToken) { _, token in
            guard token > 0 else { return }
            requestReview()
        }
    }
}

extension View {
    /// Present the system App Store review prompt whenever `manager` signals an
    /// eligible moment. Attach once, high in the view tree.
    func reviewPrompt(_ manager: ReviewRequestManager) -> some View {
        modifier(ReviewPromptModifier(manager: manager))
    }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/review/ReviewRequestManager.kt
package com.h3consultingpartners.ifatccompanion.core.review

import com.h3consultingpartners.ifatccompanion.core.platform.Clock
import com.h3consultingpartners.ifatccompanion.core.platform.KeyValueStore

/**
 * When — and whether — to ask the pilot to rate the app.
 *
 * A port of `IFATCCompanion/Review/ReviewRequestManager.swift`. Both stores apply their own
 * throttle on top of anything the app does (Play's is undocumented; Apple's is three per
 * 365 days), so a request is exactly that: a request. What this decides is whether asking is
 * *appropriate*, which is a product question and not the store's.
 *
 * The product rule is that a prompt may only appear at one of two calm moments — connected
 * and idle at the gate before the first ATC call, or arrived and parked with the flight
 * done. Never mid-flight: a pilot asked to rate the app while being vectored is being
 * interrupted at exactly the moment the app is supposed to be useful.
 *
 * This decides only. Presenting the sheet is the platform's job, and on Android that means
 * an Activity — which is why this holds no reference to one and lives in `:core`, where the
 * whole gate can be tested against a fake clock rather than by waiting three days.
 */
class ReviewRequestManager(
    private val store: KeyValueStore,
    private val clock: Clock = Clock.system,
) {

    /** The two low-workload windows a prompt is permitted in. */
    enum class Trigger {
        /** Connected and idle at the gate, before the first ATC call of the session. */
        BEFORE_FIRST_CALL,

        /** Arrived and parked — the flight is complete. */
        AFTER_FLIGHT_COMPLETE,
    }

    /**
     * Stamp the install date, once, so the install-age gate has a baseline.
     *
     * Called at start-up rather than from an `init` block: the properties it writes are
     * declared below it, and Kotlin will not let a constructor touch them. Idempotent — a
     * second call on an install that already has a date does nothing.
     */
    fun noteAppStarted() {
        if (installDateMillis == null) installDateMillis = clock.nowMillis()
    }

    /**
     * Record a fully completed flight — arrived and parked.
     *
     * Drives the engagement gate, so it must be called exactly once per completed flight. A
     * double count is not harmful in itself, but it brings the first prompt forward past the
     * point the product rule chose.
     */
    fun recordFlightCompleted() {
        completedFlights += 1
    }

    /**
     * Whether every app-side gate currently permits a prompt.
     *
     * Public rather than private because the answer is worth recording: the store's own
     * throttle means a prompt that was requested and never shown is indistinguishable from
     * one that was never requested, and on a device that is the only way to tell whether the
     * wiring works at all.
     */
    fun isEligible(): Boolean {
        if (completedFlights < MINIMUM_COMPLETED_FLIGHTS) return false
        val now = clock.nowMillis()
        // No install date means noteAppStarted has never run, which means this is the first
        // launch — the one moment Apple and Play both say not to ask in.
        val installed = installDateMillis ?: return false
        if (now - installed < MINIMUM_INSTALL_AGE_MILLIS) return false
        lastPromptMillis?.let { if (now - it < MINIMUM_INTERVAL_MILLIS) return false }
        if (promptsInLastYear(now) >= MAXIMUM_PROMPTS_PER_YEAR) return false
        return true
    }

    /**
     * Ask, if and only if every gate passes. Returns whether the caller should launch the
     * store's review flow.
     *
     * Safe to call at either allowed moment: it self-limits and is a no-op otherwise. The
     * request is recorded when it is *made*, not when a prompt appears, because the app
     * cannot find out whether one appeared — and a gate that only advanced on a confirmed
     * prompt would ask again at every eligible moment forever.
     */
    fun requestReviewIfAppropriate(trigger: Trigger): Boolean {
        if (!isEligible()) return false
        val now = clock.nowMillis()
        lastPromptMillis = now
        recentPrompts = recentPrompts.filter { now - it < ONE_YEAR_MILLIS } + now
        return true
    }

    // region Persistence
    //
    // Timestamps go through the store's Double accessor: it has no Long, and epoch millis
    // (~1.7e12) is far inside the 2^53 a Double represents exactly, so nothing is lost.

    private var installDateMillis: Long?
        get() = store.getDouble(KEY_INSTALL_DATE)?.toLong()
        set(value) = store.putDouble(KEY_INSTALL_DATE, value?.toDouble())

    /** Completed flights so far. Readable for the Diagnostics row. */
    var completedFlights: Int
        get() = store.getInt(KEY_COMPLETED_FLIGHTS) ?: 0
        private set(value) = store.putInt(KEY_COMPLETED_FLIGHTS, value)

    private var lastPromptMillis: Long?
        get() = store.getDouble(KEY_LAST_PROMPT)?.toLong()
        set(value) = store.putDouble(KEY_LAST_PROMPT, value?.toDouble())

    /**
     * When each recent prompt was asked for.
     *
     * A list rather than a count, because the cap is a *rolling* year. A count would have to
     * reset on some fixed date, and a pilot who happened to fly either side of it would be
     * asked twice as often as the rule allows.
     */
    private var recentPrompts: List<Long>
        get() = store.getString(KEY_RECENT_PROMPTS)
            ?.split(",")
            ?.mapNotNull(String::toLongOrNull)
            .orEmpty()
        set(value) = store.putString(KEY_RECENT_PROMPTS, value.joinToString(","))

    private fun promptsInLastYear(now: Long): Int =
        recentPrompts.count { now - it < ONE_YEAR_MILLIS }

    // endregion

    companion object {
        /**
         * Completed flights before the app ever asks. Keeps brand-new installs and
         * first-session pilots out of the prompt entirely — someone who has not finished a
         * flight has nothing to rate.
         */
        const val MINIMUM_COMPLETED_FLIGHTS = 3

        /**
         * How old the install must be before the first ask. Defensive: a fresh install also
         * has zero completed flights, but a restore from backup can arrive with a count.
         */
        const val MINIMUM_INSTALL_AGE_MILLIS = 3L * 86_400_000

        /** About 120 days between prompts. Well inside any store's limit, and not nagging. */
        const val MINIMUM_INTERVAL_MILLIS = 120L * 86_400_000

        /**
         * An app-side mirror of the strictest store cap, so a run of tightly spaced eligible
         * moments can never overshoot it even if the spacing rule is later relaxed.
         */
        const val MAXIMUM_PROMPTS_PER_YEAR = 3

        const val ONE_YEAR_MILLIS = 365L * 86_400_000

        // The same keys iOS uses, so the two platforms' vocabularies stay in step.
        private const val KEY_INSTALL_DATE = "review.installDate"
        private const val KEY_COMPLETED_FLIGHTS = "review.completedFlights"
        private const val KEY_LAST_PROMPT = "review.lastPromptDate"
        private const val KEY_RECENT_PROMPTS = "review.recentPromptDates"
    }
}
